### PR TITLE
feat(metrics): Add team_key_transaction fake mri transformation

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR aims at solving the issue in which the `team_key_transaction` derived operation doesn't require a `metric_mri` but because of the `MetricField` it actually requires it but without using it. This doesn't create problems until the `team_key_transaction` is used in the `orderby`. In that case because the `orderby` validation (available [here](https://github.com/getsentry/sentry/blob/fb2a714db17be29d29d5ae3735651dff3b9066b0/src/sentry/snuba/metrics/query.py#L240-L240)) requires all the fields in the `orderby` to have the same entity we cannot always pass the dummy `TransactionMRI.DURATION` to the `team_key_transaction` because if we are going to use for example the `TransactionMRI.USER_MISERY` inside of the `orderby` we are going to work with a `set` entity instead of a `distribution` entity.

There were several solutions to this problem which included the rework of some metrics layer internals, however with @ahmedetefy we came to the conclusion that we can perform an automatic dummy mri injection in the transformer instead of changing the core code. This means that whenever `team_key_transaction` is used and fed into the transformer, the following special mri should be used `e:transactions/team_key_transaction@none`. Under the hood the transformer will infer the proper mri to add to each instance of `team_key_transaction` based on the entity type inferred from the `select` fields.

### CODE REVIEW

File: src/sentry/snuba/metrics/mqb_query_transformer.py [hunk 1]
```diff
@@ -9,17 +9,23 @@
...
     MetricConditionField,
     MetricField,
     MetricGroupByField,
     MetricsQuery,
 )
+from sentry.snuba.metrics.query import OrderBy
 from sentry.snuba.metrics.query import OrderBy as MetricOrderBy
 from sentry.snuba.metrics.query_builder import FUNCTION_ALLOWLIST
 
+TEAM_KEY_TRANSACTION_FAKE_MRI = "e:custom/team_key_transaction@reserved"
```
Comment (ahmedetefy) [inline src/sentry/snuba/metrics/mqb_query_transformer.py:25]:
```suggestion
TEAM_KEY_TRANSACTION_FAKE_MRI = "e:transactions/team_key_transaction@none"
```

File: src/sentry/snuba/metrics/mqb_query_transformer.py [hunk 2]
```diff
@@ -261,4 +422,17 @@ def transform_mqb_query_to_metrics_query(query: Query) -> MetricsQuery:
...
+    (select, where, groupby, orderby) = _transform_team_key_transaction_fake_mri(
+        mq_dict["select"],
+        mq_dict["where"],
+        mq_dict["groupby"],
+        mq_dict["orderby"],
+    )
+    mq_dict["select"] = select
+    mq_dict["where"] = where
+    mq_dict["groupby"] = groupby
+    mq_dict["orderby"] = orderby
```
Comment (ahmedetefy) [inline src/sentry/snuba/metrics/mqb_query_transformer.py:436]:
```suggestion
    mq_dict.update({"select": select, "where": where, "groupby": groupby, "orderby":orderby})
```

File: src/sentry/snuba/metrics/mqb_query_transformer.py [hunk 3]
```diff
@@ -235,6 +241,161 @@ def _transform_orderby(query_orderby):
...
+                        else _recursively_compute_ingested_mri(select_field.metric_mri)
+                    )
+
+        if len(mri_types) == 1:
+            # In order to simplify the code we just set the MRI of the team_key_transaction to be equal
+            # to the one of the entity in the select in order to make sure that the "get_entity" method
+            # will always return the same result.
+            mri_to_apply = mri_types[list(mri_types.keys())[0]]
+
+    return mri_to_apply
```
Comment (ahmedetefy) [inline src/sentry/snuba/metrics/mqb_query_transformer.py:292]: Lets please replace this with the call `.get_entity()`

File: src/sentry/snuba/metrics/mqb_query_transformer.py [hunk 4]
```diff
@@ -235,6 +241,161 @@ def _transform_orderby(query_orderby):
...
+
+
+def _tmp_alias_to_none(alias):
```
Comment (ahmedetefy) [inline src/sentry/snuba/metrics/mqb_query_transformer.py:295]: Lets just set a default alias to `team_key_transaction`

File: tests/snuba/api/endpoints/test_organization_events_mep.py [hunk 5]
```diff
@@ -442,6 +442,105 @@ def test_performance_homepage_query(self):
             assert field_meta["failure_rate()"] == "percentage"
             assert field_meta["failure_count()"] == "integer"
 
+    def test_user_misery_and_team_key_sort(self):
+        self.features["organizations:use-metrics-layer"] = True
```
Comment (wmak) [inline tests/snuba/api/endpoints/test_organization_events_mep.py:446]: Can you remove this? The `OrganizationEventsMetricsEnhancedPerformanceEndpointTestWithMetricLayer` class already adds this automatically

**Review by wmak — APPROVED** [review]
1 minor comment otherwise changes to the Visibility code looks good to me

File: tests/snuba/api/endpoints/test_organization_events_mep.py (hunk 5 above)
Comment (iambriccardo) [reply tests/snuba/api/endpoints/test_organization_events_mep.py:446]: Sure! Sorry for that

File: src/sentry/snuba/metrics/mqb_query_transformer.py [hunk 6]
```diff
@@ -235,6 +240,139 @@ def _transform_orderby(query_orderby):
...
+def _derive_mri_to_apply(project_ids, select, orderby):
+    mri_dictionary = {
+        "generic_metrics_distributions": TransactionMRI.DURATION.value,
+        "generic_metrics_sets": TransactionMRI.USER.value,
+    }
+    mri_to_apply = TransactionMRI.DURATION.value
+
+    # We first check if there is an order by field that has the team_key_transaction, otherwise
+    # we just use the default mri of duration.
+    has_order_by_tkt = False
```
Comment (andriisoldatenko) [inline src/sentry/snuba/metrics/mqb_query_transformer.py:252]:
```suggestion
    has_order_by_team_key_tx = False
```

Comment (andriisoldatenko) [reply src/sentry/snuba/metrics/mqb_query_transformer.py:252]: just to be less cryptic.

File: src/sentry/snuba/metrics/mqb_query_transformer.py [hunk 7]
```diff
@@ -235,6 +240,139 @@ def _transform_orderby(query_orderby):
...
+        mri_to_apply = mri_dictionary[entities.pop()]
+
+    return mri_to_apply
+
+
+def _transform_team_key_transaction_in_select(mri_to_apply, select):
+    if select is None:
+        return select
+
+    def _select(select_field):
```
Comment (andriisoldatenko) [inline src/sentry/snuba/metrics/mqb_query_transformer.py:280]:
```suggestion
    def _select_func(select_field):
```
it's a good idea to name it with suffix, especially when you use in map call later.

File: src/sentry/snuba/metrics/mqb_query_transformer.py [hunk 8]
```diff
@@ -235,6 +240,139 @@ def _transform_orderby(query_orderby):
...
+            return MetricField(
+                op=select_field.op,
+                metric_mri=mri_to_apply,
+                params=select_field.params,
+                alias=select_field.alias,
+            )
+
+        return select_field
+
+    return list(map(_select, select))
```
Comment (andriisoldatenko) [inline src/sentry/snuba/metrics/mqb_query_transformer.py:291]:
```suggestion
    return list(map(_select_func, select))
```

File: src/sentry/snuba/metrics/mqb_query_transformer.py [hunk 9]
```diff
@@ -235,6 +240,139 @@ def _transform_orderby(query_orderby):
...
+
+
+def _transform_team_key_transaction_in_where(mri_to_apply, where):
+    if where is None:
+        return where
+
+    def _where(where_field):
```
Comment (andriisoldatenko) [inline src/sentry/snuba/metrics/mqb_query_transformer.py:298]:
```suggestion
    def _where_func(where_field):
```
or another func-style idea came to my mind:
```
_where_predicate
```
but it's up to you.

**Review by andriisoldatenko — APPROVED** [review]

**Pull request merged**